core/jvm/src/main/scala/fs2/compress.scala

```diff
@@ -75,197 +101,680 @@ object compress {
+      Stream.empty
+    }
+  }
+
+  private def _inflate[F[_]](bufferSize: Int, inflater: Inflater, crc32: Option[CRC32])(
+      implicit ev: RaiseThrowable[F]
   ): Pipe[F, Byte, Byte] =
-    _.pull.uncons.flatMap {
-      case None => Pull.done
-      case Some((hd, tl)) =>
-        val inflater = new Inflater(nowrap)
-        val buffer = new Array[Byte](bufferSize)
-        inflater.setInput(hd.toArray)
-        val result =
-          _inflate_collect(inflater, buffer, ArrayBuffer.empty).toArray
-        Pull.output(Chunk.bytes(result)) >> _inflate_stream(inflater, buffer)(ev)(tl)
+    _.pull.unconsNonEmpty.flatMap {
+      case Some((deflatedChunk, deflatedStream)) =>
+        _inflate_chunk(deflatedChunk, inflater, bufferSize.max(minBufferSize), crc32) >> _inflate_stream(
+          inflater,
+          bufferSize.max(minBufferSize),
+          crc32
+        )(ev)(deflatedStream)
+      case None =>
+        Pull.done
     }.stream
 
-  private def _inflate_stream[F[_]](inflater: Inflater, buffer: Array[Byte])(
-      implicit ev: RaiseThrowable[F]
-  ): Stream[F, Byte] => Pull[F, Byte, Unit] =
-    _.pull.uncons.flatMap {
-      case Some((hd, tl)) =>
-        inflater.setInput(hd.toArray)
-        val result =
-          _inflate_collect(inflater, buffer, ArrayBuffer.empty).toArray
-        Pull.output(Chunk.bytes(result)) >> _inflate_stream(inflater, buffer)(ev)(tl)
+  private def _inflate_chunk[F[_]](
+      chunk: Chunk[Byte],
+      inflater: Inflater,
+      inflatedBufferSize: Int,
+      crc32: Option[CRC32]
+  ): Pull[F, Byte, Unit] = {
+    val bytesChunk = chunk.toBytes
+    inflater.setInput(
+      bytesChunk.values,
+      bytesChunk.offset,
+      bytesChunk.length
+    )
+    def inflateInto(inflatedBuffer: Array[Byte]): Int =
+      if (inflater.finished()) -2
+      else if (inflater.needsInput()) -1
+      else {
+        val byteCount = inflater.inflate(inflatedBuffer)
+        crc32.foreach(_.update(inflatedBuffer, 0, byteCount))
+        byteCount
+      }
+
+    def pull(): Pull[F, Byte, Unit] = {
+      val inflatedBuffer = new Array[Byte](inflatedBufferSize)
+
+      inflateInto(inflatedBuffer) match {
+        case inflatedBytes if inflatedBytes <= -2 =>
+          Pull.output(chunk)
+        case inflatedBytes if inflatedBytes == -1 =>
+          Pull.done
+        case inflatedBytes if inflatedBytes < inflatedBuffer.length =>
+          if (inflater.finished()) {
+            inflater.getRemaining match {
+              case bytesRemaining if bytesRemaining > 0 =>
+                Pull.output(asChunkBytes(inflatedBuffer, inflatedBytes)) >>
+                  Pull.output(
+                    Chunk.Bytes(
+                      bytesChunk.values,
+                      bytesChunk.offset + bytesChunk.length - bytesRemaining,
+                      bytesRemaining
+                    )
+                  ) >> Pull.done
```

Comment:
Shouldn't need the `>> Pull.done` or below

Reply:
Agreed. Both removed. Thank you.